### ISSUE TITLE
Any scope for allowing users to select/specify an image underpinning their container?

### ISSUE THREAD

**Issue by Analect** [jupyterhub/jupyterhub-deploy-docker#25]
Excuse the length of this, but it didn't make sense to break it up.

It may be out of scope here, but just wanted to ask ... or perhaps get pointed to other implementations.  I also posted an issue [here](https://github.com/ryanlovett/imagespawner/issues/2) against Ryan Lovett's [imagespawner repo](https://github.com/ryanlovett/imagespawner), having heard him mention it during the [jupyterhub-2016-workshop](https://github.com/jupyterhub/jupyterhub-2016-workshop).

My use-case is as follows:
1. users A and B have a private gitlab instance running, each with a collection of private and shared repos
2. user A wants to be able to login to a running jupyterhub deployment to interact with one (or maybe more) of his private repos ideally exposing [jupyterlab](https://github.com/jupyter/docker-stacks/wiki/Docker-recipes#run-jupyter-lab).
3. the assumption is that the instance running the jupyterhub would already have all the latest images of user A and user B's private repos.
4. those private repos would include something akin to a Dockerfile or requirements.txt defining the appropriate environment for a notebook to work (correct packages etc..) ... not dissimilar to how binder works.
5. the repo file-system (as captured in the docker image), would pre-populate a file-explorer widget in the jupyterlab exposed in the container launched for a user based on the files as captured in the repo when the image was generated.
6. any changes the user makes to the repo from that jupyterlab, is captured in a docker volume (as per this repo's architecture) .. and therefore could theoretically be used to commit changes back to the gitlab instance.
7. let's say user A wants to work separately but simultaneously on another of his gitlab repos ... is there scope for him to launch more than one containerised environment while logged into jupyterhub as the same user ... but access another container based on a different repo image?

From [here](https://github.com/willingc/jupyterhub-deploy-docker/blob/c89e4ec0d0db0e03f7c0a51a0336f9d313b59487/README.md#spawner-prepare-the-jupyter-notebook-image), I know different images can be used to underpin a jupyterhub, but changing requires a hub [restart](https://github.com/willingc/jupyterhub-deploy-docker/blob/c89e4ec0d0db0e03f7c0a51a0336f9d313b59487/README.md#if-i-change-the-name-of-the-notebook-server-image-to-spawn-do-i-need-to-restart-jupyterhub) and it's not clear to me if it's possible to have different containers spawned from different images running on the same hub?

Would something like the imagespawner, linked to above, facilitate multiple containers based on different images running on the same jupyterhub?

If so, how about the scenario where the same user spins up different containers based on different images?

For step 4 above, is it enough that a private repo contains the [singleuser script](https://github.com/jupyter/docker-stacks/wiki/Docker-recipes#use-with-jupyterhubs-dockerspawner) for it to be compatible with jupyterhub ... or is there something else that qualifies an image for use on jupyterhub, as per the [docker-stacks](https://github.com/jupyter/docker-stacks) images?

I know that some of this is bordering more on the 'binder' space, and specifically their [custom deployments](http://docs.mybinder.org/custom-deployments), although that brings with it greater complexities in terms of managing a kubernetes cluster, and I wanted to exhaust the possibilities with just using docker for now and the existing tooling around jupyterhub.

I'm aware there are initiatives afoot elsewhere, such as this [one](https://github.com/jupyterhub/kubespawner/pull/11) using a dynamic volumes to be associated with spawned user environments in kubernetes, and maybe ultimately the scenario I present above with a user being able to interact with multiple repos in containerised jupyterlab environments requires something like kubernetes for it to be achieved?

Thanks, as ever, to those infinitely better informed about this stuff than I am.

**Comment by jtyberg** [comment]
@Analect I believe I understand your scenario.  Allow me to set aside some of the details for a moment and focus on a few key questions.

> let's say user A wants to work separately but simultaneously on another of his gitlab repos ... is there scope for him to launch more than one containerised environment while logged into jupyterhub as the same user

I'm not sure if JupyterHub currently supports multiple containers (running on multiple IPs/ports) for a single user.   In this reference implementation, there are two components that assume a single container per user: the [configurable-http-proxy](https://github.com/jupyterhub/jupyterhub/blob/70727c4940d0c9c1ee88da4ee6dbb380025a7d35/jupyterhub/orm.py#L117) (routes user to container IP, port) and the [DockerSpawner](https://github.com/jupyterhub/dockerspawner/blob/master/dockerspawner/dockerspawner.py#L75) (maps user to container ID).  I am not an expert on the proxy, nor do I know if it can support multiple targets per user, but I know the DockerSpawner would require some changes.

> I know different images can be used to underpin a jupyterhub, but changing requires a hub restart and it's not clear to me if it's possible to have different containers spawned from different images running on the same hub?

JupyterHub provides hooks for the spawner to allow the user to customize their deployment, which might include selecting from a list of available images.  See the [spawner options form](https://github.com/jupyterhub/jupyterhub/blob/798346dbe8c9492f32cb7ceea7c6dffe635a85e4/docs/source/spawners.md#spawner-options-form).    The [imagespawner](https://github.com/ryanlovett/imagespawner) is actually an example that leverages this capability.

>  is it enough that a private repo contains the singleuser script for it to be compatible with jupyterhub...

The single-user script launches a special-purpose notebook server that is aware of the hub.  Again, I'm no expert, but this script handles all the setup to ensure only a single user can authenticate to the server.  When a user's notebook server is spawned, the single-user script must be invoked to ensure proper security setup.

> and maybe ultimately the scenario I present above with a user being able to interact with multiple repos in containerised jupyterlab environments requires something like kubernetes for it to be achieved?

I believe your instinct here is correct.  Kubernetes [supports multiple containers per pod](http://kubernetes.io/docs/user-guide/pods/multi-container/), although it is unclear to me whether the containers can be spawned from different images.  If so, then the kubespawner would also have to support the mapping of a single user to the multiple containers in the pod.  Even still, something tells me that JupyterHub might need changes as well.

In summary, custom deployments based on user options should be well within reach.  However, support for multiple simultaneous containers per user is likely to require some code refactoring.  

/cc @minrk @yuvipanda

**Comment by minrk** [comment]
As @jtyberg suggested, an individual user cannot currently have multiple *running* servers per user. But through the spawner form, as illustrated in [imagespawner](https://github.com/ryanlovett/imagespawner), or [everware](http://everware.xyz), you can switch your image with:

1. stop current server
2. select new image
3. start new server

without restarting the Hub. Only the single-user server needs restarting, as long as the options are exposed.

It is on the roadmap to support multiple servers per user, but there's a bit of work to do to scrub out the current `user.server` singleton assumptions.

Still another option would be to have a *single* server container for your server, and expose each project environment via kernelspecs. That way each notebook can be in a different env without restarting the notebook server.

**Comment by Analect** [comment]
Thank you @jtyberg and @minrk for your quick and detailed response.
Just as a quick follow-up @minrk  ... for the steps to stop current server, select new image and start new server ... are these possible to achieve with a series of REST calls to the [jupyterhub API](http://petstore.swagger.io/?url=https://raw.githubusercontent.com/jupyter/jupyterhub/master/docs/rest-api.yml#/default) ... or is the API not necessarily conceived for that use-case?

**Comment by minrk** [comment]
You can do this via the REST API:

1. DELETE /api/users/:name/server
2. POST /api/users/:name/server with `options` in the JSON body

Though I was expecting humans to do it via regular pages:

1. control panel -> stop my server
2. start my server -> gets form via `options_form` for selecting new server type

**Comment by Analect** [comment]
Just so I'm clear, an `options` example, as implemented by imagespawner would be [this](https://github.com/ryanlovett/imagespawner/blob/master/imagespawner.py#L45-L68) ... which could be modified to point at a [list of images](https://docker-py.readthedocs.io/en/latest/api/#images) available on the host.  The [start](https://github.com/ryanlovett/imagespawner/blob/master/imagespawner.py#L71-L80) is then effectively what gets called in `POST /api/users/:name/server` above ... or does there need to be some other mechanism to link that API end-point to that `start` function?

**Comment by minrk** [comment]
Yes, that's right. `options_form` presents users with a form to select some inputs (e.g. a dropdown list of images) before starting the server. You can provide the same options when sending a POST request directly. In both cases, the options are passed as `self.user_options` in Spawner.start. In the form case, the form data passes through `Spawner.options_from_form` to produce an options dict, whereas the POST request passes the options through unmodified from the JSON body of the request.

**Comment by Analect** [comment]
Thanks @minrk
